File: 2024/src/typescript/day3/p2.ts
import { parseInput } from "../utils/readFile";

const filePath = '/home/daniel/Documents/repos/adventofcode/2024/inputs/day3.txt';

async function solution() {
  const file = await parseInput(filePath, true);
  if (typeof file !== 'string') {
    throw new Error('Expected a string');
  }

  let state = 'enabled';
  const validMulInstructions: string[] = [];

  const regex = /don't\(\)|\b\w*do\b|mul\(\d+,\d+\)*\)/g;
  const matches = file.matchAll(regex);
  const numRegex = /\d+,\d+/g;

  for (const match of matches) {
    const instr = match[0];

    if (instr === "don't()") {
      state = 'disabled'
    } else if (/^\w*do$/.test(instr)) {
      state = 'enabled';
    } else if (instr.startsWith('mul(')) {
      if (state == 'enabled') {
        validMulInstructions.push(instr)
      }
    }
  }

  let sum = 0

  for (const instr of validMulInstructions){
    const nums = instr.match(numRegex)
    if (nums) {
      const itemsToMult = nums[0]
        const [first, second] = itemsToMult.split(',')
        const action = Number(first) * Number(second)
        sum += action
    }
  }

  console.log(sum)
}

solution()

File: 2024/src/python/day4/p2.py
from ..utils.readInputs import parse_input

filePath = '/home/daniel/Documents/repos/adventofcode/2024/inputs/day4.txt'
file = parse_input(filePath)

rows = len(file)
cols = len(file[0])

def check_sequence(seq: str):
  return seq == 'SAM' or seq == 'MAS'


sum = 0

for  i in range(rows):
  for j in range(cols):
    if i <= rows - 3 and j <= cols - 3 and (file[i][j] == 'M' or file[i][j] == 'S') and (file[i][j+2] == 'M' or file[i][j+2] == 'S'):
      ltr_diag = ''.join(file[i+k][j+k] for k in range(3))
      rtl_diag = ''.join(file[i+k][j+2-k] for k in range(3))
      if check_sequence(ltr_diag) and check_sequence(rtl_diag):
        sum += 1

      
print(sum)